docs/source/spans.rst

```diff
@@ -0,0 +1,145 @@
+Spans
+=====
+.. currentmodule:: distributed
+
+.. note::
+
+    This is an experimental feature and may rapidly change without a deprecation cycle.
+
+Dask offers various diagnostics and :doc:`fine-performance-metrics` about tasks,
+grouping them by their prefix (commonly, the name of the function being called). This
+can be suboptimal:
+
+- your client code may be sophisticated enough that correlating lines on the client and
+  tasks being run on the cluster may prove difficult;
+- the same function may be applied to different parts of your workflow, with a different
+  performance profile;
+- you may be in a multitenancy setup, where part of the load on the cluster was not
+  originated by your client code.
+
+In these situations, it may be useful to attach arbitrary tags to your workflow, or
+segments of it.
+To do so, you should use the :func:`span` context manager inside the client code.
+
+For example:
+
+.. code-block:: python
+
+    import dask.config
+    import dask.dataframe as dd
+    from distributed import Client, span
+
+    dask.config.set({"optimization.fuse.active": False})
+    client = Client()
+
+    @span("ML preprocessing")
+    def preprocess(df):
+        ...
+        return df
+
+    @span("Model training")
+    def train(df):
+        ...
+        return model
+
+    with span("Alice's workflow"):
+        with span("data load"):
+            df = dd.read_parquet(...)
+
+        df = preprocess(df)
+        model = train(df)
+
+    model = model.compute()
+
+In the above example, note how the :func:`span` context manager is used as a decorator
+too, and how it can be nested.
+The example will create the following spans on the scheduler:
+
+- ``("Alice's workflow", )``
+- ``("Alice's workflow", "data load")``
+- ``("Alice's workflow", "ML preprocessing")``
+- ``("Alice's workflow", "Model training")``
+
+Each of the spans will be mapped to the tasks matching the segment of the graph that was
+defined inside its context manager. The top-level span will be mapped to all tasks of
```

Comment:
Is it really only the top-level span that will be mapped to tasks of its children or _any_ parent span?

Reply:
fixed